Use raw-dylib for Windows synchronization functions

ChrisDenton opened the pull request:
Fixes #123999 by using the raw-dylib feature to specify the DLL to load the Windows futex functions from (e.g. [`WaitOnAddress`](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress)). This avoids reliance on the import library causing that issue.

With apologies to @bjorn3, as it's currently necessary to revert this for cranelift.

ChrisDenton commented:
Oh silly bors, this is really a libs thing. Just with a bit of cranelift sprinkled on top,

r? libs

bjorn3 commented on library/std/src/sys/pal/windows/c.rs:
```diff
@@ -356,18 +356,34 @@ compat_fn_with_fallback! {
+// Use raw-dylib to import synchronization functions to workaround issues with the older mingw import library.
+cfg_if::cfg_if! {
+if #[cfg(not(target_vendor = "win7"))] {
+    #[cfg(target_arch = "x86")]
+    #[link(name = "api-ms-win-core-synch-l1-2-0", kind = "raw-dylib", import_name_type = "undecorated")]
```
Instead of duplicating the extern block, maybe use `#[cfg_attr]` for `#[link]`?

ChrisDenton committed:
Use raw-dylib for Windows futex APIs

This is a workaround for older mingw `synchronization` import library not working on at least some system.

ChrisDenton committed:
Cranelift: Revert raw-dylib for Windows futex APIs

bors commented:
:pushpin: Commit 5b024d6550c5c43848386df52f455da13ca79dea has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 5b024d6550c5c43848386df52f455da13ca79dea -->
<!-- homu: {"type":"Approved","sha":"5b024d6550c5c43848386df52f455da13ca79dea","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

ChrisDenton commented:
I'm nominating this for a beta backport, even though this isn't technically a Rust issue. See linked issue for more but the short version is that the windows-2019 github action runner comes with a years old mingw that has a bug which will prevent windows-gnu std from working if compiled on that machine. This PR works around the bug. It's a small PR so I think it's worth it to make sure that stable works out of the box on windows-2019.

Note that the windows-latest runner isn't affected.
